composer/core/state.py

```diff
@@ -468,13 +531,32 @@ def __init__(
         self.profiler: Optional[Profiler] = None
 
         self.deepspeed_config = deepspeed_config
-        self.fsdp_config = fsdp_config
-        self.fsdp_auto_wrap = fsdp_auto_wrap
+        parallelism_config = parallelism_config or {}
+        self.fsdp_config = parallelism_config.get('fsdp', None)
+        self.tp_config = parallelism_config.get('tp', None)
+        if self.fsdp_config is not None:
+            # Add an earlier call to patch_pytorch as we require device_mesh slicing before any
+            # model wrapping.
+            from composer.trainer.mosaic_fsdp import patch_pytorch
+
+            patch_pytorch()
+
+        if self.tp_config is not None:
+            if version.parse(torch.__version__.split('.dev')[0]) < version.parse('2.3.0'):
+                raise ValueError('Tensor parallelism (TP) requires torch>=2.3.0.')
+            if self.fsdp_config is None:
+                raise ValueError(
+                    'Tensor parallelism (TP) currently requires FSDP to be enabled. '
+                    'An empty `fsdp_config` can be specified to enable FSDP with '
+                    'default settings.',
+                )
 
         if self.load_monolith_rank0_only:
-            assert fsdp_config is not None
+            if self.tp_config is not None:
+                raise ValueError('load_fsdp_monolith_rank0_only is not compatible with tensor parallelism (TP).')
```

Comment:
Is there a fix the user should use if they hit this?

Reply:
We should include unsupported edge cases in docs -- would help decrease support burden and we can always edit down the line